### ISSUE TITLE
Doesn't work with cluster autoscaler

### ISSUE THREAD

**Issue by rahulinux** [neondatabase/autoscaling#64]
## Steps to reproduce

- `autoscaler-scheduler` takes `resource.limits=4Gi` also into consideration when scheduling pods which resulted to use existing nodes. because of this `cluster-autoscaler` never kicks in to add more nodes and pods get stuck in pending.  

- Tested same with default scheduler and it works. in default scheduler case it does `preempt` existing pods (lower priority dummy pods by overprovisioning). 

## Expected result

[From CA FAQ](https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/FAQ.md#how-does-scale-up-work):

>   Whenever a Kubernetes scheduler fails to find a place to run a pod, it sets "schedulable" PodCondition to false and reason to "unschedulable". If there are any items in the unschedulable pods list, Cluster Autoscaler tries to find a new place to run them.

- Should set pod to `unschedulable` condition 
- Preempt low priority pods. 

## Actual result

```
cluster-autoscaler-76fdd8b4d-tb49v cluster-autoscaler I0223 15:23:21.970141       1 filter_out_schedulable.go:118] Pod default.compute-node-ep-curly-snowflake-209252-66c8479797-jxrtm marked as unschedulable can be scheduled on node ip-10-6-93-124.us-east-2.compute.internal (based on
 hinting). Ignoring in scale up.
 
cluster-autoscaler-76fdd8b4d-tb49v cluster-autoscaler I0223 15:21:51.479724       1 expendable.go:41] Pod compute-node-ep-falling-sound-830915-5b687b957-5vsvb will be scheduled after low priority pods are preempted on ip-10-6-93-124.us-east-2.compute.internal. Ignoring in scale up.
```

## Environment

- dev-us-east-2-beta

## Logs, links

<details>

```
cluster-autoscaler-76fdd8b4d-tb49v cluster-autoscaler I0223 15:21:51.479724       1 expendable.go:41] Pod compute-node-ep-falling-sound-830915-5b687b957-5vsvb will be scheduled after low priority pods are preempted on ip-10-6-93-124.us-east-2.compute.internal. Ignoring in scale up.
cluster-autoscaler-76fdd8b4d-tb49v cluster-autoscaler I0223 15:21:51.479735       1 expendable.go:41] Pod compute-node-ep-bold-wave-096020-6f474cbcfc-bnndw will be scheduled after low priority pods are preempted on ip-10-6-93-124.us-east-2.compute.internal. Ignoring in scale up.
cluster-autoscaler-76fdd8b4d-tb49v cluster-autoscaler I0223 15:21:51.479740       1 expendable.go:41] Pod compute-node-ep-delicate-unit-803343-5f8c76bcfc-gbplk will be scheduled after low priority pods are preempted on ip-10-6-93-124.us-east-2.compute.internal. Ignoring in scale up.
cluster-autoscaler-76fdd8b4d-tb49v cluster-autoscaler I0223 15:21:51.479745       1 expendable.go:41] Pod compute-node-ep-floral-forest-176446-6c559cfdd-jkzs7 will be scheduled after low priority pods are preempted on ip-10-6-93-124.us-east-2.compute.internal. Ignoring in scale up.
cluster-autoscaler-76fdd8b4d-tb49v cluster-autoscaler I0223 15:21:51.479749       1 expendable.go:41] Pod compute-node-ep-silent-hall-359947-9d8868cb9-nfdvd will be scheduled after low priority pods are preempted on ip-10-6-93-124.us-east-2.compute.internal. Ignoring in scale up.
cluster-autoscaler-76fdd8b4d-tb49v cluster-autoscaler I0223 15:21:51.479753       1 expendable.go:41] Pod compute-node-ep-old-fire-020831-58787f4496-q7f5n will be scheduled after low priority pods are preempted on ip-10-6-93-124.us-east-2.compute.internal. Ignoring in scale up.
cluster-autoscaler-76fdd8b4d-tb49v cluster-autoscaler I0223 15:21:51.479764       1 expendable.go:41] Pod compute-node-ep-curly-rice-649198-674f4c4595-mvqbf will be scheduled after low priority pods are preempted on ip-10-6-93-124.us-east-2.compute.internal. Ignoring in scale up.
cluster-autoscaler-76fdd8b4d-tb49v cluster-autoscaler I0223 15:21:51.479769       1 expendable.go:41] Pod compute-node-ep-frosty-credit-432032-5f4769cddb-qz6kq will be scheduled after low priority pods are preempted on ip-10-6-93-124.us-east-2.compute.internal. Ignoring in scale up.

cluster-autoscaler-76fdd8b4d-tb49v cluster-autoscaler I0223 15:21:51.479837       1 filter_out_schedulable.go:157] Pod default.compute-node-ep-frosty-credit-407552-84488c4986-cvtjd marked as unschedulable can be scheduled on node ip-10-6-22-73.us-east-2.compute.internal. Ignoring in scale up.
cluster-autoscaler-76fdd8b4d-tb49v cluster-autoscaler I0223 15:21:51.479859       1 filter_out_schedulable.go:157] Pod default.compute-node-ep-curly-snowflake-209252-dd796f684-mmqxb marked as unschedulable can be scheduled on node ip-10-6-60-247.us-east-2.compute.internal. Ignoring in scale up.
cluster-autoscaler-76fdd8b4d-tb49v cluster-autoscaler I0223 15:21:51.479878       1 filter_out_schedulable.go:157] Pod default.compute-node-ep-rough-mouse-294138-667b7c548c-kzzvg marked as unschedulable can be scheduled on node ip-10-6-22-73.us-east-2.compute.internal. Ignoring in scale up.
cluster-autoscaler-76fdd8b4d-tb49v cluster-autoscaler I0223 15:21:51.479885       1 filter_out_schedulable.go:170] 0 pods were kept as unschedulable based on caching
cluster-autoscaler-76fdd8b4d-tb49v cluster-autoscaler I0223 15:21:51.479888       1 filter_out_schedulable.go:171] 3 pods marked as unschedulable can be scheduled.
cluster-autoscaler-76fdd8b4d-tb49v cluster-autoscaler I0223 15:21:51.479895       1 filter_out_schedulable.go:79] Schedulable pods present
cluster-autoscaler-76fdd8b4d-tb49v cluster-autoscaler I0223 15:21:51.479915       1 static_autoscaler.go:419] No unschedulable pods


cluster-autoscaler-76fdd8b4d-tb49v cluster-autoscaler I0223 15:23:21.970098       1 filter_out_schedulable.go:65] Filtering out schedulables
cluster-autoscaler-76fdd8b4d-tb49v cluster-autoscaler I0223 15:23:21.970141       1 filter_out_schedulable.go:118] Pod default.compute-node-ep-curly-snowflake-209252-66c8479797-jxrtm marked as unschedulable can be scheduled on node ip-10-6-93-124.us-east-2.compute.internal (based on hinting). Ignoring in scale up.
cluster-autoscaler-76fdd8b4d-tb49v cluster-autoscaler I0223 15:23:21.970159       1 filter_out_schedulable.go:118] Pod default.compute-node-ep-floral-forest-176446-f4fdb5ccf-5x282 marked as unschedulable can be scheduled on node ip-10-6-77-255.us-east-2.compute.internal (based on hinting). Ignoring in scale up.
cluster-autoscaler-76fdd8b4d-tb49v cluster-autoscaler I0223 15:23:21.970189       1 filter_out_schedulable.go:118] Pod default.compute-node-ep-bold-wave-096020-fff6c68d6-xhxfl marked as unschedulable can be scheduled on node ip-10-6-93-124.us-east-2.compute.internal (based on hinting). Ignoring in scale up.
cluster-autoscaler-76fdd8b4d-tb49v cluster-autoscaler I0223 15:23:21.970196       1 filter_out_schedulable.go:132] Filtered out 3 pods using hints
cluster-autoscaler-76fdd8b4d-tb49v cluster-autoscaler I0223 15:23:21.970218       1 filter_out_schedulable.go:157] Pod default.compute-node-ep-delicate-unit-803343-68695455f6-ht5xs marked as unschedulable can be scheduled on node ip-10-6-49-94.us-east-2.compute.internal. Ignoring in scale up.
cluster-autoscaler-76fdd8b4d-tb49v cluster-autoscaler I0223 15:23:21.970237       1 filter_out_schedulable.go:157] Pod default.compute-node-ep-old-fire-020831-85dd49ddb8-tmjnq marked as unschedulable can be scheduled on node ip-10-6-93-124.us-east-2.compute.internal. Ignoring in scale up.
cluster-autoscaler-76fdd8b4d-tb49v cluster-autoscaler I0223 15:23:21.970257       1 filter_out_schedulable.go:157] Pod default.compute-node-ep-frosty-credit-407552-d6c495bbf-b22cm marked as unschedulable can be scheduled on node ip-10-6-29-162.us-east-2.compute.internal. Ignoring in scale up.
cluster-autoscaler-76fdd8b4d-tb49v cluster-autoscaler I0223 15:23:21.970276       1 filter_out_schedulable.go:157] Pod default.compute-node-ep-falling-sound-830915-fc49c5458-jdv84 marked as unschedulable can be scheduled on node ip-10-6-29-162.us-east-2.compute.internal. Ignoring in scale up.
cluster-autoscaler-76fdd8b4d-tb49v cluster-autoscaler I0223 15:23:21.970294       1 filter_out_schedulable.go:157] Pod default.compute-node-ep-silent-hall-359947-6454f5c944-2zrwv marked as unschedulable can be scheduled on node ip-10-6-49-94.us-east-2.compute.internal. Ignoring in scale up.
cluster-autoscaler-76fdd8b4d-tb49v cluster-autoscaler I0223 15:23:21.970313       1 filter_out_schedulable.go:157] Pod default.compute-node-ep-frosty-credit-432032-57f9f55c54-8dphp marked as unschedulable can be scheduled on node ip-10-6-60-247.us-east-2.compute.internal. Ignoring in scale up.
cluster-autoscaler-76fdd8b4d-tb49v cluster-autoscaler I0223 15:23:21.970331       1 filter_out_schedulable.go:157] Pod default.compute-node-ep-curly-rice-649198-6c6dcb48ff-fgmts marked as unschedulable can be scheduled on node ip-10-6-29-162.us-east-2.compute.internal. Ignoring in scale up.
cluster-autoscaler-76fdd8b4d-tb49v cluster-autoscaler I0223 15:23:21.970337       1 filter_out_schedulable.go:170] 0 pods were kept as unschedulable based on caching
cluster-autoscaler-76fdd8b4d-tb49v cluster-autoscaler I0223 15:23:21.970342       1 filter_out_schedulable.go:171] 10 pods marked as unschedulable can be scheduled.
cluster-autoscaler-76fdd8b4d-tb49v cluster-autoscaler I0223 15:23:21.970347       1 filter_out_schedulable.go:79] Schedulable pods present
cluster-autoscaler-76fdd8b4d-tb49v cluster-autoscaler I0223 15:23:21.970358       1 static_autoscaler.go:419] No unschedulable pods
cluster-autoscaler-76fdd8b4d-tb49v cluster-autoscaler I0223 15:23:21.970369       1 static_autoscaler.go:466] Calculating unneeded nodes
```

</details>

**Comment by rahulinux** [comment]
It's running and throwing following error:

<details>
```
autoscale-scheduler-868c55c896-lq2gn kube-second-scheduler 	vCPU: node reserved 1 -> 1, node other resources 1 -> 1 rounded (810m -> 1080m raw, 420m margin)
autoscale-scheduler-868c55c896-lq2gn kube-second-scheduler 	 mem: node reserved 3 -> 4, node other resources 3 -> 4 slots (3132Mi -> 4176Mi raw, 947252Ki margin)
autoscale-scheduler-868c55c896-lq2gn kube-second-scheduler I0228 09:41:11.726028       1 run.go:50] [autoscale-enforcer] Received autoscaler-agent request (client = 10.6.14.16:52572) {ProtoVersion:v1.1 Pod:default:compute-node-ep-calm-sky-909249-z9mxs Resources:{VCPU:1 Mem:4} Metrics:<nil>}
autoscale-scheduler-868c55c896-lq2gn kube-second-scheduler I0228 09:41:11.726068       1 run.go:237] [autoscale-enforcer] Updating pod default:compute-node-ep-calm-sky-909249-z9mxs metrics <nil> -> <nil>
autoscale-scheduler-868c55c896-lq2gn kube-second-scheduler 2023/02/28 09:41:11 http: panic serving 10.6.14.16:52572: runtime error: invalid memory address or nil pointer dereference
autoscale-scheduler-868c55c896-lq2gn kube-second-scheduler goroutine 2315399806 [running]:
autoscale-scheduler-868c55c896-lq2gn kube-second-scheduler net/http.(*conn).serve.func1()
autoscale-scheduler-868c55c896-lq2gn kube-second-scheduler 	/usr/local/go/src/net/http/server.go:1850 +0xbf
autoscale-scheduler-868c55c896-lq2gn kube-second-scheduler panic({0x1c1b300, 0x31d9db0})
autoscale-scheduler-868c55c896-lq2gn kube-second-scheduler 	/usr/local/go/src/runtime/panic.go:890 +0x262
autoscale-scheduler-868c55c896-lq2gn kube-second-scheduler github.com/neondatabase/autoscaling/pkg/plugin.(*podState).isBetterMigrationTarget(...)
autoscale-scheduler-868c55c896-lq2gn kube-second-scheduler 	/workspace/pkg/plugin/state.go:646
autoscale-scheduler-868c55c896-lq2gn kube-second-scheduler github.com/neondatabase/autoscaling/pkg/plugin.migrationQueue.Less(...)
autoscale-scheduler-868c55c896-lq2gn kube-second-scheduler 	/workspace/pkg/plugin/queue.go:43
autoscale-scheduler-868c55c896-lq2gn kube-second-scheduler container/heap.up({0x2215700, 0xc000318580}, 0xc000317340?)
autoscale-scheduler-868c55c896-lq2gn kube-second-scheduler 	/usr/local/go/src/container/heap/heap.go:92 +0x7e
autoscale-scheduler-868c55c896-lq2gn kube-second-scheduler container/heap.Push({0x2215700, 0xc000318580}, {0x1c766a0?, 0xc000317340?})
autoscale-scheduler-868c55c896-lq2gn kube-second-scheduler 	/usr/local/go/src/container/heap/heap.go:53 +0x5a
autoscale-scheduler-868c55c896-lq2gn kube-second-scheduler github.com/neondatabase/autoscaling/pkg/plugin.(*migrationQueue).addOrUpdate(0xc001706540?, 0x3?)
autoscale-scheduler-868c55c896-lq2gn kube-second-scheduler 	/workspace/pkg/plugin/queue.go:17 +0x37
autoscale-scheduler-868c55c896-lq2gn kube-second-scheduler github.com/neondatabase/autoscaling/pkg/plugin.(*AutoscaleEnforcer).updateMetricsAndCheckMustMigrate(0x1bf3f80?, 0xc000317340, 0xc0003184b0, 0x0)
autoscale-scheduler-868c55c896-lq2gn kube-second-scheduler 	/workspace/pkg/plugin/run.go:244 +0x1f5
autoscale-scheduler-868c55c896-lq2gn kube-second-scheduler github.com/neondatabase/autoscaling/pkg/plugin.(*AutoscaleEnforcer).handleAgentRequest(0xc0008a66e0, {0x2, {{0xc000efa600, 0x25}, {0xc00272c870, 0x7}}, {0x1, 0x4}, 0x0})
autoscale-scheduler-868c55c896-lq2gn kube-second-scheduler 	/workspace/pkg/plugin/run.go:152 +0x1fc
autoscale-scheduler-868c55c896-lq2gn kube-second-scheduler github.com/neondatabase/autoscaling/pkg/plugin.(*AutoscaleEnforcer).runPermitHandler.func1({0x2212ec0, 0xc0012a20e0}, 0xc000ddf100)
autoscale-scheduler-868c55c896-lq2gn kube-second-scheduler 	/workspace/pkg/plugin/run.go:54 +0x5f4
autoscale-scheduler-868c55c896-lq2gn kube-second-scheduler net/http.HandlerFunc.ServeHTTP(0xc0012a20e0?, {0x2212ec0?, 0xc0012a20e0?}, 0x1f0fc6e?)
autoscale-scheduler-868c55c896-lq2gn kube-second-scheduler 	/usr/local/go/src/net/http/server.go:2109 +0x2f
autoscale-scheduler-868c55c896-lq2gn kube-second-scheduler net/http.(*ServeMux).ServeHTTP(0x0?, {0x2212ec0, 0xc0012a20e0}, 0xc000ddf100)
autoscale-scheduler-868c55c896-lq2gn kube-second-scheduler 	/usr/local/go/src/net/http/server.go:2487 +0x149
autoscale-scheduler-868c55c896-lq2gn kube-second-scheduler net/http.serverHandler.ServeHTTP({0x22026f8?}, {0x2212ec0, 0xc0012a20e0}, 0xc000ddf100)
autoscale-scheduler-868c55c896-lq2gn kube-second-scheduler 	/usr/local/go/src/net/http/server.go:2947 +0x30c
autoscale-scheduler-868c55c896-lq2gn kube-second-scheduler net/http.(*conn).serve(0xc001280780, {0x22142b8, 0xc0005be180})
autoscale-scheduler-868c55c896-lq2gn kube-second-scheduler 	/usr/local/go/src/net/http/server.go:1991 +0x607
autoscale-scheduler-868c55c896-lq2gn kube-second-scheduler created by net/http.(*Server).Serve
autoscale-scheduler-868c55c896-lq2gn kube-second-scheduler 	/usr/local/go/src/net/http/server.go:3102 +0x4db
autoscale-scheduler-868c55c896-lq2gn kube-second-scheduler I0228 09:41:13.028192       1 watch.go:51] [autoscale-enforcer] watch: Received delete event for pod default:compute-node-ep-spring-night-877498-fdcbb5686-x2wbm
```
</details>

it's working as expected: evicting low priority pods and CA triggers scaleUp action.

**Comment by sharnoff** [comment]
The panic should be fixed by #67. I suspect the _behavior_ described in this issue is resolved, but I worry it'll be much harder to fix the broader issue of "cluster-autoscaler expects its view of cluster resources is accurate". Would be good to discuss.

**Comment by sharnoff** [comment]
Closing this issue because the immediate behavior is fixed. Opened #77 to track / discuss the longer-term VMs + cluster-autoscaler issue we might have to deal with.
